content/articles/how-to-build-a-web-crawler-with-java/index.md

```diff
@@ -0,0 +1,178 @@
+    }
+
+    public void crawl(String rootURL, int breakpoint){
+        urlQueue.add(rootURL);
+        visitedURLs.add(rootURL);
+
+        while(!urlQueue.isEmpty()){
+
+            // remove the next url string from the queue to begin traverse.
+            String s = urlQueue.remove();
+            String rawHTML = "";
+            try{
+                // create url with the string.
+                URL url = new URL(s);
+                BufferedReader in = new BufferedReader(new InputStreamReader(url.openStream()));
+                String inputLine = in.readLine();
+
+                // read every line of the HTML content in the URL
+                // and concatenate each line to the rawHTML string until every line is read.
+                while(inputLine  != null){
+                    rawHTML += inputLine;
+
+                    inputLine = in.readLine();
+                }
+                in.close();
+            }catch (Exception e){
+                e.printStackTrace();
+            }
+
+
+            // create a regex pattern matching a URL
+            // that will validate the content of HTML in search of a URL.
+            String urlPattern = "(www|http:|https:)+[^\s]+[\\w]";
+            Pattern pattern = Pattern.compile(urlPattern);
+            Matcher matcher = pattern.matcher(rawHTML);
+
+            // Each time the regex matches a URL in the HTML,
+            // add it to the queue for the next traverse and to the list of visited URLs.
+            while(matcher.find()){
+                String actualURL = matcher.group();
+
+                if(!visitedURLs.contains(actualURL)){
+                    visitedURLs.add(actualURL);
+                    System.out.println("Website found with URL " + actualURL);
+                    urlQueue.add(actualURL);
+                }
+                // exit the inner loop when it reaches the breakpoint.
+                if(breakpoint == 0){
+                    break;
+                }
+                breakpoint--;
+            }
+
+            // exit the outermost loop when it reaches the breakpoint.
+            if(breakpoint == 0){
+                break;
+            }
+        }
+    }
+}
+
+
+
+```
+
+In the `crawl` method of the above code snippet, the `rootURL` is the starting point of the crawler and the `breakpoint` represents how many URLs you want your crawler to discover. 
+
+The algorithm starts by adding the root URL to the queue and the list of visited URLs. Then it reads the HTML content of the URL using the `BufferedReader` API to search for any embedded URLs. It uses the specified regex pattern to discover URLs in the HTML. If you’re new to regex, [here](https://www.section.io/engineering-education/regex-implementation-with-java/) is a tutorial I created that will get you started on implementing regex with Java.
+
+These operations will iterate until the crawler has discovered the number of URLs specified in your `breakpoint`.
```

Comment:
You could have specified as these constraints before you start implementing.

Thus, making the reader aware of these constraints.

Specify the limit that you've set

Reply:
I already specified the limit in the paragraph just after the `main` method code block, before the conclusion.

Reply:
I've added a Prerequisites heading.

Reply:
> You must mention it before you start with the code. How do you expect the reader to read until last to understand something you've written at the top?
> 
> Please mention all such parameters before hand. It makes it easier for the reader to understand.

Isn't that what the Pseudocode heading I have there is meant for?